libs/editor/egui_editor/src/input/mutation.rs

```diff
@@ -177,193 +203,276 @@ pub fn calc(
         Modification::Indent { deindent } => {
             // if we're in a list item, tab/shift+tab will indent/de-indent
             // otherwise, tab will insert a tab and shift tab will do nothing
-            let galley_idx = galleys.galley_at_char(current_cursor.selection.1);
-            let galley = &galleys[galley_idx];
-            if let Some(annotation) = &galley.annotation {
-                match annotation {
-                    Annotation::Item(item_type, indent_level) => {
-                        // todo: this needs more attention e.g. list items doubly indented using 2-space indents
-                        let galley_text = &buffer[(galley.range.start(), galley.range.end())];
-                        let indent_seq = if galley_text.starts_with('\t') {
-                            "\t"
-                        } else if galley_text.starts_with("    ") {
-                            "    "
-                        } else if galley_text.starts_with("  ") {
-                            "  "
-                        } else {
-                            "\t"
-                        };
+            let mut processed_galleys = HashSet::new();
+            let mut indented_galleys = HashMap::new();
+            let mut renumbered_galleys = HashMap::new();
+
+            let ast_text_ranges = bounds.ast.find_intersecting(current_cursor.selection, true);
+
+            // iterate forwards for indent and backwards for de-indent because when indenting, the indentation of the
+            // prior item constraints the indentation of the current item, and when de-indenting, the indentation of
+            // the next item constraints the indentation of the current item
```

Comment:
renumbering list items for multiple (de-)indentations in one frame is much more complicated than I think users realize, please appreciate me 🥺